// printer-open-api/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RetCode = exports.TaskPrintState = exports.PrinterStatus = exports.ApiName = void 0;
var ApiName;
(function (ApiName) {
    ApiName["BIND_PRINTERS"] = "Open_printerAddlist";
    ApiName["PRINT"] = "Open_printMsg";
    ApiName["UNBIND_PRINTERS"] = "Open_printerDelList";
    ApiName["TASK_STATUS"] = "Open_queryOrderState";
    ApiName["PRINTER_STATUS"] = "Open_queryPrinterStatus";
    ApiName["QUERY_PRINTERS"] = "Open_queryPrinterList";
    ApiName["QUERY_TASKS"] = "Open_queryTasksList";
    ApiName["CLEAR_PENDING_TASKS"] = "Open_delPrinterSqs";
})(ApiName = exports.ApiName || (exports.ApiName = {}));
;
var PrinterStatus;
(function (PrinterStatus) {
    PrinterStatus["OFFLINE"] = "\u79BB\u7EBF";
    PrinterStatus["ONELINE_WORKING"] = "\u5728\u7EBF\uFF0C\u5DE5\u4F5C\u72B6\u6001\u6B63\u5E38";
    PrinterStatus["ONLINE_WITH_EXCEPTION"] = "\u5728\u7EBF\uFF0C\u5DE5\u4F5C\u72B6\u6001\u4E0D\u6B63\u5E38\u3002";
})(PrinterStatus = exports.PrinterStatus || (exports.PrinterStatus = {}));
;
var TaskPrintState;
(function (TaskPrintState) {
    TaskPrintState[TaskPrintState["PRINTED"] = 1] = "PRINTED";
    TaskPrintState[TaskPrintState["NOT_PRINTED"] = 0] = "NOT_PRINTED";
})(TaskPrintState = exports.TaskPrintState || (exports.TaskPrintState = {}));
;
var RetCode;
(function (RetCode) {
    RetCode[RetCode["SUCCESS"] = 0] = "SUCCESS";
    RetCode[RetCode["INVALID_SIG_OR_TIMESTAMP"] = -3] = "INVALID_SIG_OR_TIMESTAMP";
    RetCode[RetCode["EMPTY_REQUEST"] = -2] = "EMPTY_REQUEST";
    RetCode[RetCode["WRONG_REQUEST_HEADER"] = -2] = "WRONG_REQUEST_HEADER";
    RetCode[RetCode["UNMATCHED_PRINTER_USER_RELATION"] = 1002] = "UNMATCHED_PRINTER_USER_RELATION";
    RetCode[RetCode["INVALID_REQUEST"] = 1001] = "INVALID_REQUEST";
})(RetCode = exports.RetCode || (exports.RetCode = {}));
